### PR TITLE
Feat/update most used tags

### PR DESCRIPTION
**Pull request by J3suis**
- Add a ways to see all mostusedtags history by user with routes /tagging/mostusedtags-history
- Link most used tags to /tagging/compliance response

### CODE REVIEW

File: tagging/routes/routeGetMostUsedTags.go [hunk 1]
```diff
@@ -54,3 +55,37 @@ func routeGetMostUsedTags(r *http.Request, a routes.Arguments) (int, interface{}
...
 }
+
+func routeGetMostUsedTagsHistory(r *http.Request, a routes.Arguments) (int, interface{}) {
+	logger := jsonlog.LoggerFromContextOrDefault(r.Context())
+	u := a[users.AuthenticatedUser].(users.User)
+	tx := a[db.Transaction].(*sql.Tx)
+	dbRes, err := models.MostUsedTagsHistoryByUser(tx, u.Id)
+	if err != nil {
+		logger.Error("Could not fetch most used tags.", err.Error())
+		return http.StatusInternalServerError, nil
```
Comment (GabrielRavier) [inline tagging/routes/routeGetMostUsedTags.go]: From what I can see in other code, you can also return the details of the error as the second return value is an `interface{}` and something like `error` will be handled perfectly well, as can be seen from code in `users/create.go`, for example.

File: tagging/routes/routeGetMostUsedTags.go [hunk 2]
```diff
@@ -54,3 +55,37 @@ func routeGetMostUsedTags(r *http.Request, a routes.Arguments) (int, interface{}
...
+	}
+
+	if dbRes == nil {
+		return http.StatusInternalServerError, map[string]interface{}{
```
Comment (GabrielRavier) [inline tagging/routes/routeGetMostUsedTags.go]: Same here, you don't have to return a `map[string]interface{}`, you can just go and return something like `errors.New("No history reports available")` (unless the `map[string]interface{}` looks much better or something like that)

File: tagging/routes/routes.go [hunk 3]
```diff
@@ -56,6 +56,17 @@ func init() {
...
 	}.H().Register("/tagging/mostusedtags")
 
+	routes.MethodMuxer{
+		http.MethodGet: routes.H(routeGetMostUsedTagsHistory).With(
+			db.RequestTransaction{db.Db},
+			users.RequireAuthenticatedUser{users.ViewerAsParent},
+			routes.Documentation{
+				Summary:     "get most used tags",
+				Description: "Responds with most used tags for a user.",
+			},
```
Comment (GabrielRavier) [inline tagging/routes/routes.go:66]: This documentation seems inaccurate, considering it's a copy-paste of the one from `/tagging/mostusedtags`

**Review by GabrielRavier — CHANGES_REQUESTED** [review]

**Review by GabrielRavier — APPROVED** [review]
LGTM now

File: tagging/routes/routeGetMostUsedTags.go [hunk 4]
```diff
@@ -54,3 +56,35 @@ func routeGetMostUsedTags(r *http.Request, a routes.Arguments) (int, interface{}
...
+	u := a[users.AuthenticatedUser].(users.User)
+	tx := a[db.Transaction].(*sql.Tx)
+	dbRes, err := models.MostUsedTagsHistoryByUser(tx, u.Id)
+	if err != nil {
+		logger.Error("Could not fetch most used tags.", err.Error())
+		return http.StatusInternalServerError, err.Error()
+	}
+
+	if dbRes == nil {
+		return http.StatusInternalServerError, errors.New("no history reports available")
```
Comment (simonmeyerrr) [inline tagging/routes/routeGetMostUsedTags.go]: Should be a `404` error

File: tagging/routes/routeGetMostUsedTags.go [hunk 5]
```diff
@@ -54,3 +56,35 @@ func routeGetMostUsedTags(r *http.Request, a routes.Arguments) (int, interface{}
...
+	}
+	history := make([]map[string]interface{}, 0)
+	for _, mut := range dbRes {
+		tagsList := []string{}
+		err = json.Unmarshal([]byte(mut.Tags), &tagsList)
+		if err != nil {
+			logger.Error("Could not unmarshal most used tags.", err.Error())
+			return http.StatusInternalServerError, err
```
Comment (simonmeyerrr) [inline tagging/routes/routeGetMostUsedTags.go]: Use `continue` so a little error does not block the user to get valid history data.

File: tagging/routes/routeGetMostUsedTags.go [hunk 6]
```diff
@@ -54,3 +56,35 @@ func routeGetMostUsedTags(r *http.Request, a routes.Arguments) (int, interface{}
...
+		}
+		if mut == nil || len(tagsList) == 0 {
```
Comment (simonmeyerrr) [inline tagging/routes/routeGetMostUsedTags.go]: It's to late to check if `mut` is nil as you already use it earlier

**Review by simonmeyerrr — CHANGES_REQUESTED** [review]

**Review by simonmeyerrr — APPROVED** [review]

**Pull request merged**